### PR TITLE
feat: create cancellation API

### PR DESCRIPTION
**Pull request by udyding**
## Notion ticket link

[Ticket Name](https://www.notion.so/uwblueprintexecs/Tasks-38b49f83f6ce4b07ae188dabb37631e8?p=63fbd360033049debb58fcd3a6c7a699)

## Implementation description
* Added `camper/cancel/:chargeId` route allowing for cancellation of registration for all campers under the transaction's charge ID
* This endpoint will first check if eligible for cancellation: must be cancelling >= 30 days before start of camp, there are campers with the charge ID, the camp exists

## Steps to test
1. Test manually in Postman - create a PUT request to the endpoint with a charge ID
2. Test multiple cases: if there are no campers with the charge ID, camp that is associated with the campers being deleted doesn't exist, test all edge cases for the 30 day eligibility for deletion, then ensure that for a valid deletion that the campers from the camp are deleted and the camper documents are also deleted
3. Test rolling back the camp's camper deletion if deleting the campers throws an error
4. Note that I already tested all these cases ^

## What should reviewers focus on?
* Question: will the `dates` field on the `Camp` be sorted by first date of camp to last date? Needed as the first date of the dates field is checked for eligibility

## Checklist
- [x] My PR name is descriptive and in imperative tense
- [x] My commit messages are descriptive and in imperative tense. My commits are atomic and trivial commits are squashed or fixup'd into non-trivial commits
- [x] I have run the appropriate linter(s)
- [x] I have requested a review from the PL, as well as other devs who have background knowledge on this PR or who will be building on top of this PR

### CODE REVIEW

**Review by NirmalHegde — APPROVED** [review]
lgtm

File: backend/typescript/services/interfaces/camperService.ts [hunk 1]
```diff
@@ -23,6 +23,14 @@ interface ICamperService {
    * @throws Error if camper retrieval fails
    */
   getCampersByCampId(campId: string): Promise<Array<CamperDTO>>;
+
+  /**
```
Comment (amanda-guo) [inline backend/typescript/services/interfaces/camperService.ts:27]: should these comments specify deleting a camper?

**Comment by amanda-guo** [comment]
to address the question about the`dates` field in `Camp` - not sure how it's stored, but would it make sense to include the start date as a separate param? to guarantee that you have the right date? or is that too redundant?

File: backend/typescript/services/implementations/camperService.ts [hunk 2]
```diff
@@ -167,6 +167,77 @@ class CamperService implements ICamperService {
...
+      const numericalChargeId = Number(chargeId);
+      const campers: Array<Camper> = await MgCamper.find({
+        chargeId: numericalChargeId,
+      });
+
+      if (!campers.length) {
+        throw new Error(`Campers with charge ID ${chargeId} not found.`);
+      }
+
+      const camp: Camp | null = await MgCamp.findById(campers[0].camp);
```
Comment (amanda-guo) [inline backend/typescript/services/implementations/camperService.ts:182]: just confirming, multiple campers with the same chargeId must all be registered for the same camp?

**Comment by amanda-guo** [comment]
tested some basic cases and seems to work!

**Comment by udyding** [comment]
> to address the question about the`dates` field in `Camp` - not sure how it's stored, but would it make sense to include the start date as a separate field in the Camp model? to guarantee that you have the right date? or is that too redundant?

Start date has been added to the camp field in another PR i believe, resolved :)

File: backend/typescript/rest/camperRoutes.ts [hunk 3]
```diff
@@ -66,4 +66,14 @@ camperRouter.get("/", async (req, res) => {
   }
 });
 
+/* Delete all campers with the chargeId with */
+camperRouter.put("/cancel/:chargeId", async (req, res) => {
```
Comment (annechung) [inline backend/typescript/rest/camperRoutes.ts:70]: dkm can we change this to a delete endpoint, thinking about it again I think that makes more sense
```suggestion
camperRouter.delete("/cancel/:chargeId", async (req, res) => {
```

File: backend/typescript/services/implementations/camperService.ts [hunk 4]
```diff
@@ -127,6 +127,77 @@ class CamperService implements ICamperService {
 
     return camperDtos;
   }
+
+  async deleteCampersByChargeId(chargeId: string): Promise<void> {
+    try {
+      const numericalChargeId = Number(chargeId);
```
Comment (annechung) [inline backend/typescript/services/implementations/camperService.ts:133]: I just checked the Stripe docs and chargeId is actually a string D: can we change the type to a string in the camper model and remove this?

File: backend/typescript/services/implementations/camperService.ts [hunk 5]
```diff
@@ -127,6 +127,77 @@ class CamperService implements ICamperService {
...
+      );
+
+      if (daysUntilStartOfCamp < 30) {
+        throw new Error(
+          `Campers' camp with campId ${campers[0].camp} has a start date in less than 30 days.`,
+        );
+      }
+
+      const camperIds = campers.map((camper) => camper.id);
+      const camperObjectIds = [...camp.campers]; // clone the full array of campers for rollback
```
Comment (annechung) [inline backend/typescript/services/implementations/camperService.ts:165]: nit: can we rename camperObjectIds to oldCamperIds?

**Review by annechung — COMMENTED** [review]
looks good, just a few small comments

**Review by annechung — APPROVED** [review]
tested and looks good!! feel free to merge it in :D
